crux-lucene/src/crux/lucene.clj

```diff
@@ -10,19 +10,54 @@
   (:import crux.query.VarBinding
            java.io.Closeable
            java.nio.file.Path
+           java.time.Duration
+           java.util.Date
            org.apache.lucene.analysis.Analyzer
            org.apache.lucene.analysis.standard.StandardAnalyzer
            [org.apache.lucene.document Document Field$Store StoredField StringField TextField]
            [org.apache.lucene.index DirectoryReader IndexWriter IndexWriterConfig Term]
            org.apache.lucene.queries.function.FunctionScoreQuery
            org.apache.lucene.queryparser.classic.QueryParser
            [org.apache.lucene.search BooleanClause$Occur BooleanQuery$Builder DoubleValuesSource IndexSearcher Query ScoreDoc SearcherManager TermQuery TopDocs]
-           [org.apache.lucene.store Directory FSDirectory]))
+           [org.apache.lucene.store Directory FSDirectory]
+           [java.util.concurrent Future Executors TimeUnit]))
 
-(defrecord LuceneNode [directory analyzer index-writer searcher-manager indexer]
+(defn debounce [name f wait-ms]
+  (let [!last-run-time (atom (.getTime (Date.)))
+        !is-scheduled? (atom false)
+        executor (Executors/newSingleThreadScheduledExecutor (cio/thread-factory name))
+        schedule-run! (fn run []
+                        (when (false? @!is-scheduled?)
+                          (let [now (.getTime (Date.))
+                                next-run (+ @!last-run-time wait-ms)]
+                            (if (< next-run now)
+                              (do 
+                                (reset! !last-run-time (.getTime (Date.)))
+                                (f))
+                              (reset! !is-scheduled?
+                                      (.schedule executor
+                                                 ^Runnable
+                                                 (fn []
+                                                   (reset! !is-scheduled? false)
+                                                   (reset! !last-run-time (.getTime (Date.)))
+                                                   (f))
+                                                 ^long (- next-run now)
+                                                 TimeUnit/MILLISECONDS))))))]
+    [(if (= wait-ms 0) f schedule-run!)
+     (reify
+       Closeable
+       (close [_]
+         (when-not (false? @!is-scheduled?)
```

Comment:
if there is an fsync scheduled when you're being asked to close, would you want to call it immediately?